### DISCUSSION TITLE
Expanding export functionality

### DISCUSSION THREAD

**Discussion by mikeemm** [raysan5/raylib#4524] [Features]
Hi there, wasn't sure if this is something that should be proposed in a PR, so I thought I'd float the idea around here first.

Raylib currently supports loading models in various formats (OBJ, GLTF, and a few others). However, exporting only supports OBJ through ExportMesh(), and while said function is made to be able to be extended with other formats, it is restricted in a number of ways:

- It cannot save material data (seemingly not skeleton data either, though I haven't looked well at how animations work with Raylib yet, so I won't assume).
- It doesn't allow for passing any export options.

For OBJ, this means we can't pair exported meshes with an MTL file, which would also need to be generated separately. When implementing GLTF/GLB, it not only means we won't be able to include images in them, but even if we did, we also couldn't let the user choose the way data is to be included (embedded buffers or stored separately as .bin and .png/jpg). I don't know how it is for .m3d, .iqm and .vox, but I'd bet at least one of those will also be affected by these restrictions one way or another once they are implemented.

I think the way to get over these restrictions would be to...
1) Separate each format export process into their own function (currently just OBJ, so ExportOBJ(), but in the future also others like ExportGLTF(), etc. ), each with their potential options passed as arguments.
2) Change ExportMesh() to call these functions, with default options.
3) Create a new function ExportModel() to also call these functions with default options, but with access to variables from Model so it could also include material data.

Below's a rough idea of what this would look like (please excuse my tragic mockery of UML):

![proposal](https://github.com/user-attachments/assets/b2f43262-8993-4a06-a14f-f459303f30af)

I personally need to implement more export functionality for the program I am working on, so I thought I might as well offer to build it directly in Raylib. This would however require adding new functions and restructuring an existing one, so I thought it'd probably be best to share my idea and get feedback before starting work on it. Obviously @raysan5 but I also want to see what other people think about it all. I understand model exporting is not a commonly used feature, but I feel that given the use of Raylib not just for videogame development but also for the creation of other kinds of graphical applications, it could greatly help those making programs in any way related to 3D model editing.

Anyway thanks for reading, have a great day!

**Comment by raysan5** [comment]
@mikeemm Thanks for the nice diagram and description of the proposal, `ExportMesh()` was jsut added for convenience and adding support for other 3d formats would require a considerable amount of work and new functionality and I don't think many users will need. So, for now I prefer to keep it limited as it is. Users with more advance needs can import the specific exporters at user-side, there is full acces to all the `Model` data.

**Reply by mikeemm** [reply]
@raysan5 Thanks for responding, I apologize for my late reply.

I still believe these additions would be worth the effort. While it is not necessary nor the only reason to use the library, applications using Raylib will feel compelled to use its `Model` structure together with its loading functionality. Similarly, at least from my experience, applications that play with models and edit their structure might still want to use Raylib's `Model` for rendering even while implementing their own model structure holding a more malleable representation of data.

I feel that it is then a missed opportunity to force users to make their own exporters for runtime-optimized formats such as glTF or OBJ when they may already be using Raylib's `Model`, and could all benefit from a standardized and safe method to export models without having to reinvent the wheel for their particular use cases despite already using Raylib's data structures. Raylib also already includes library dependencies for object loading such as `cgltf`, which if I am not mistaken also includes exporting functionality and would need to be unnecessarily included in the user's code too if they instead had to create their own export functions using it.

I have been focusing on other aspects of development for the moment, but eventually I will need to get back to creating exporters, and for my personal use case I will be creating said exporters based on Raylib's `Model` anyway. When I can finally spend some time on this, you think I could maybe send a PR for `ExportModel()` for you to check out, showcasing OBJ exporting expanded to work together with MTL data? I don't think the result of that would add much baggage to Raylib at all, and that way you could see if you liked the way it came out. If so, then that could be potentially expanded in the future with glTF/M3D/VOX implementations - in the worst of cases, OBJs would now be able to exported together with `Model`'s material data, and that would be it.

**Reply by raysan5** [reply]
@mikeemm Still thinking that most users probably don't need it BUT if the exporters could be self-contained on its own functions similar to their counter-parts `LoadOBJ()`, `LoadIQM()`, `LoadGLTF()`, like `SaveOBJ()`, `SaveIQM()`, `SaveGLTF()`, returning byte memory buffers to be exported with `ExportModel()`, in that case it could be added.

**Reply by mikeemm** [reply]
@raysan5 Yeah that was what I had in mind. All export functions would be made separately and could be called on their own; `ExportModel()` would just call different Save...() functions inside of it depending on the desired filename's format, much like what `LoadModel()` currently does. `ExportMesh()` would now also use these new save functions, except it would call them with a "blank" model created from the selected mesh; also no MTL export when exporting OBJs from a single mesh, obviously.
